### ISSUE TITLE
2.0-65 Linux Appimage Use Motion not available on Gamepad input with Nintendo Switch Pro controller SDL input

### ISSUE THREAD

**Issue by theofficialgman** [cemu-project/Cemu#1097]
### Current Behavior

use motion is greyed out on `Nintendo Switch Pro Controller [SDLController]` when the wii u gamepad input is selected

### Expected Behavior

expect that use motion is available

### Steps to Reproduce

use latest appimage with nintendo switch pro controller. attempt to use SDL input and select use motion but it is greyed out

**Comment by theofficialgman** [comment]
same issue on flatpak

**Comment by Squall-Leonhart** [comment]
insufficient information to act on or investigate.

What controller is this.
what linux distribution and version

**Comment by theofficialgman** [comment]
> What controller is this. 

Controller name already provided twice. To reiterate, it's an official Nintendo switch pro controller.

> what linux distribution and version

Also already partially provided. Flatpak distribution and appimage on Ubuntu Mantic 23.10.
I tested the appimages on another system on Ubuntu Noble 24.04 development and do not have this issue.

> Your github profile tells me you should know better than to post such a vague and unactionable issue.

I typically don't provide repetitive and extra irrelevant information. At the time it appeared that distro was likey irrelevant as long as hid-nintendo driver was being used from the kernel as sdl is provided in the appimage statically linked in the main binary.

**Comment by Squall-Leonhart** [comment]
6.5.0.25.25 has a proposed(and accepted) fix for switch pro controller issues.

**Comment by theofficialgman** [comment]
I looked into this. it appears that one system is in the `input` group (the one that works) and the other is not. the system with the user in the `input` group can read all `/dev/input/` nodes without root. the one that isn't cannot and can only read the device node for `Nintendo Switch Pro Controller` and not `Nintendo Switch Pro Controller IMU` because only the former has extended permissions for my user (getfacl shows this).

**Comment by Squall-Leonhart** [comment]
Then this SDL issue explains whats up https://github.com/libsdl-org/SDL/issues/8125

**Comment by Squall-Leonhart** [comment]
Actually, apologies but its this one https://github.com/libsdl-org/SDL/issues/6478

The current dkms-nintendo driver only supplies the motion data as a dev/input/event

**Comment by theofficialgman** [comment]
> Then this SDL issue explains whats up [libsdl-org/SDL#8125](https://github.com/libsdl-org/SDL/issues/8125)
> Actually, apologies but its this one [libsdl-org/SDL#6478](https://github.com/libsdl-org/SDL/issues/6478)

**Comment by theofficialgman** [comment]
> Then this SDL issue explains whats up [libsdl-org/SDL#8125](https://github.com/libsdl-org/SDL/issues/8125)
> Actually, apologies but its this one [libsdl-org/SDL#6478](https://github.com/libsdl-org/SDL/issues/6478)

no and no. the problem is not that it is exposed separated. that is solved in SDL2 years ago now (see the second issue PR close reason). the problem is that `Nintendo Switch Pro Controller IMU` doesn't have extended permissions to be read by my user without `input` group access.

**Comment by Squall-Leonhart** [comment]
https://github.com/libsdl-org/SDL/pull/7697 was first added to SDL 3.0 to partially (not fully) mitigate
the udev rule with read access is still required to get the sensor.

https://github.com/libsdl-org/SDL/pull/8336 backports the change to SDL 2.30 which has only just come out, should in part correct this issue and #753.

**Comment by theofficialgman** [comment]
> [libsdl-org/SDL#7697](https://github.com/libsdl-org/SDL/pull/7697) was first added to SDL 3.0 to partially (not fully) mitigate the udev rule with read access is still required to get the sensor.
> 
> [libsdl-org/SDL#8336](https://github.com/libsdl-org/SDL/pull/8336) backports the change to SDL 2.30 which has only just come out, should in part correct this issue and #753.

I don't see what that PR changes. This already works on SDL2.28.5 (what is used in latest CEMU 2.0-65) as long as read access exists. I am not really sure why it works already since that PR is not in 2.28.5

The PR to systemd linked within that PR (https://github.com/libsdl-org/SDL/pull/7697#issuecomment-1593369338) is what is needed though to add uaccess so that regular users can read the IMU https://github.com/systemd/systemd/pull/22860

**Comment by theofficialgman** [comment]
even after removing the user from the `input` group and rebooting it still works on my noble machine. it appears that on connecting the controller in CEMU the controller restarts or reconnects (the lights change). according to the PRs hidraw based input and motion should already be functioning in SDL2 so I guess that is what is being used here (instead of hid-nintendo event# inputs). It works on my one system through HIDRAW as a side affect of having Steam installed, which installs a udev rule for hidraw access by user as part of the package

`/lib/udev/rules.d/60-steam-input.rules`
```
# Nintendo Switch Pro Controller over USB hidraw
KERNEL=="hidraw*", ATTRS{idVendor}=="057e", ATTRS{idProduct}=="2009", MODE="0660", TAG+="uaccess"

# Nintendo Switch Pro Controller over bluetooth hidraw
KERNEL=="hidraw*", KERNELS=="*057E:2009*", MODE="0660", TAG+="uaccess"
```

so basically, the only way it is going to work in the current SDL2 is if the user has that udev rule provided for hidraw access via user, like steam provides.

I have confirmed adding that rule (not all of steam, just the rule) to the system that previously was not working fixes it.

**Comment by theofficialgman** [comment]
testing the gyro in Nintendo Land it works but it appears to be offset by 90deg (I am looking at the ground with the controller flat, I have to hold the controller up/down to make it look forward).

Not sure if this is incorrect parsing by SDL2 of the hidraw input or incorrect parsing by Cemu.

**Comment by AlphaCraft9658** [comment]
This happens with any SDL Controller on my side.

**Comment by mttdr** [comment]
Hello, I confirm I had this same issue (unable to enable motion controls), but with a dualshock 4 controller, connected over bluetooth if that matters.

Fresh install of Cemu flatpak on Fedora 41.

https://github.com/cemu-project/Cemu/issues/1097#issuecomment-1952626670 did fix it and also made motion controls available in `antimicrox`.

**Comment by AlphaCraft9658** [comment]
> This still happens with any SDL Controller on my side. The option for motion is grayed out and I have a controller with a gyro. It works elsewhere, but not here. So the udev rule doesn't fix it without modifications. Can this be fixed in Cemu or otherwise? It should work out of the box.

Nevermind. The mode I was using the controller in doesn't support Gyro input. It's an XInput limitation. Switch mode works.

**Comment by inmymum** [comment]
On the latest version of arch with cemu from the aur, even using the rule gyro is still greyed out.

On a side note if I run cemu with sudo the system can in fact access gyro, so it is definetly a permissons thing.

If someone can help would be apprieciated.

**Comment by YogiGemini** [comment]
> On the latest version of arch with cemu from the aur, even using the rule gyro is still greyed out. Controller is also a pro controller with the same usb uuid of `057e:2009`
> 
> On a side note if I run cemu with sudo the system can in fact access gyro, so it is definetly a permissons thing.
> 
> If someone can help would be apprieciated.

I had the same problem as you on Ubuntu 24.04.
Here's how I solved it:

- Download [joycond](https://github.com/DanielOgorchock/joycond) following the installation instructions.

- Edit the file located in `/lib/udev/rules.d/89-joycond.rules` as administrator and replace all `MODE="0600"` values with `MODE="0666"`.
The file should then look like this:

```
# Keep steam from accessing hidraw for pro controller
# Nintendo Switch Pro Controller over USB hidraw
KERNEL=="hidraw*", ATTRS{idVendor}=="057e", ATTRS{idProduct}=="2009", MODE="0666", TAG-="uaccess", RUN+="/bin/setfacl -b /dev/%k"
# Nintendo Switch Pro Controller over bluetooth hidraw
KERNEL=="hidraw*", KERNELS=="*057E:2009*", MODE="0666", TAG-="uaccess", RUN+="/bin/setfacl -b /dev/%k"


ACTION!="add", GOTO="joycond_end"
SUBSYSTEM!="input", GOTO="joycond_end"
KERNEL!="event*", GOTO="joycond_end"
ATTRS{id/vendor}=="057e", ATTRS{id/product}=="2009", ATTRS{name}!="*Combined*", ATTRS{name}!="*Virtual*", ATTRS{name}!="*IMU*", TAG+="joycond", MODE="0666"
ATTRS{id/vendor}=="057e", ATTRS{id/product}=="2006", ATTRS{name}!="*Combined*", ATTRS{name}!="*Virtual*", ATTRS{name}!="*IMU*", TAG+="joycond", MODE="0666"
ATTRS{id/vendor}=="057e", ATTRS{id/product}=="2007", ATTRS{name}!="*Combined*", ATTRS{name}!="*Virtual*", ATTRS{name}!="*IMU*", TAG+="joycond", MODE="0666"
ATTRS{id/vendor}=="057e", ATTRS{id/product}=="200e", ATTRS{name}!="*Combined*", ATTRS{name}!="*Virtual*", ATTRS{name}!="*IMU*", TAG+="joycond", MODE="0666"
ATTRS{id/vendor}=="057e", ATTRS{id/product}=="2017", ATTRS{name}!="*Combined*", ATTRS{name}!="*Virtual*", ATTRS{name}!="*IMU*", TAG+="joycond", MODE="0666"
ATTRS{id/vendor}=="057e", ATTRS{id/product}=="2019", ATTRS{name}!="*Combined*", ATTRS{name}!="*Virtual*", ATTRS{name}!="*IMU*", TAG+="joycond", MODE="0666"

LABEL="joycond_end"
```

- If you have the file located in `/lib/udev/rules.d/60-steam-input.rules` mentioned by @theofficialgman, it seems that joycond conflicts with it.
So I cancelled the rule lines with `#` (also as admin) like this:

```
# Nintendo Switch Pro Controller over USB hidraw
#KERNEL=="hidraw*", ATTRS{idVendor}=="057e", ATTRS{idProduct}=="2009", MODE="0660", TAG+="uaccess"

# Nintendo Switch Pro Controller over bluetooth hidraw
#KERNEL=="hidraw*", KERNELS=="*057E:2009*", MODE="0660", TAG+="uaccess"
```

- Finally, follow the joycond usage instructions to connect your controller correctly 👍 

This worked for me, on the latest version of Cemu (2.5).
I hope you find this useful =)

**Comment by inmymum** [comment]
For some clarification, I don't have steam installed, just used the steam udev rules. I haven't tested this in a while. Just before I ask the aur maintainer to add this udev rule I want to verify it actually works

**Comment by rafal-tech** [comment]
> even after removing the user from the `input` group and rebooting it still works on my noble machine. it appears that on connecting the controller in CEMU the controller restarts or reconnects (the lights change). according to the PRs hidraw based input and motion should already be functioning in SDL2 and I have found that is what is being used here (instead of hid-nintendo event# inputs).
> 
> It works on my one system through HIDRAW as a side affect of having Steam installed, which installs a udev rule for hidraw access by user as part of the package
> 
> `/lib/udev/rules.d/60-steam-input.rules`
> 
> ```
> # Nintendo Switch Pro Controller over USB hidraw
> KERNEL=="hidraw*", ATTRS{idVendor}=="057e", ATTRS{idProduct}=="2009", MODE="0660", TAG+="uaccess"
> 
> # Nintendo Switch Pro Controller over bluetooth hidraw
> KERNEL=="hidraw*", KERNELS=="*057E:2009*", MODE="0660", TAG+="uaccess"
> ```
> 
> [60-steam-input.rules.zip](https://github.com/cemu-project/Cemu/files/14333958/60-steam-input.rules.zip)
> 
> so basically, the only way it is going to work in the current SDL2 (2.28.5) is if the user has that udev rule provided for hidraw access via user, like steam provides.
> 
> I have confirmed adding that rule (not all of steam, just the rule) to the system that previously was not working fixes it.

Hello so its enough to put this file in the folder ? Or should i do something else ? Its not working on my lego s

**Comment by inmymum** [comment]
> > even after removing the user from the `input` group and rebooting it still works on my noble machine. it appears that on connecting the controller in CEMU the controller restarts or reconnects (the lights change). according to the PRs hidraw based input and motion should already be functioning in SDL2 and I have found that is what is being used here (instead of hid-nintendo event# inputs).
> > It works on my one system through HIDRAW as a side affect of having Steam installed, which installs a udev rule for hidraw access by user as part of the package
> > `/lib/udev/rules.d/60-steam-input.rules`
> > ```
> > # Nintendo Switch Pro Controller over USB hidraw
> > KERNEL=="hidraw*", ATTRS{idVendor}=="057e", ATTRS{idProduct}=="2009", MODE="0660", TAG+="uaccess"
> > 
> > # Nintendo Switch Pro Controller over bluetooth hidraw
> > KERNEL=="hidraw*", KERNELS=="*057E:2009*", MODE="0660", TAG+="uaccess"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [60-steam-input.rules.zip](https://github.com/cemu-project/Cemu/files/14333958/60-steam-input.rules.zip)
> > so basically, the only way it is going to work in the current SDL2 (2.28.5) is if the user has that udev rule provided for hidraw access via user, like steam provides.
> > I have confirmed adding that rule (not all of steam, just the rule) to the system that previously was not working fixes it.
> 
> Hello so its enough to put this file in the folder ? Or should i do something else ? Its not working on my lego s

Last I checked this didn't work on fedora or arch.
